be/ut/dtm0_log_ut.c

```diff
@@ -0,0 +1,298 @@
+		m0_buf_free(&temp_buf);
+	} else {
+		rc = (m0_buf_is_set(buf) == 0);
+	}
+
+	return rc;
+}
+
+int ut_dl_init(struct m0_dtm0_tx_desc *txd, struct m0_buf *buf, int rand)
+{
+	int rc;
+
+	rc = ut_dl_init_txd(txd, rand);
+	if (rc != 0)
+		return rc;
+
+	rc = ut_dl_init_buf(buf, rand);
+	if (rc != 0)
+		m0_dtm0_tx_desc_fini(txd);
+
+	return rc;
+}
+
+void ut_dl_fini(struct m0_dtm0_tx_desc *txd, struct m0_buf *buf)
+{
+	ut_dl_fini_txd(txd);
+	ut_dl_fini_buf(buf);
+}
+
+bool ut_dl_verify_log_rec(struct m0_dtm0_log_rec *rec, int rand)
+{
+	return ut_dl_verify_buf(&rec->dlr_pyld, rand) &&
+               ut_dl_verify_txd(&rec->dlr_txd, rand);
+}
+
+
+void test_volatile_dtm0_log(void)
+{
+	int                     i;
+	int                     rc;
+	struct m0_dtm0_clk_src  cs;
+	struct m0_dtm0_tx_desc  txd[UT_DTM0_LOG_MAX_LOG_REC];
+	struct m0_buf           buf[UT_DTM0_LOG_MAX_LOG_REC] = {};
+	struct m0_buf           empty_buf                    = {};
+	struct m0_dtm0_log_rec *rec[UT_DTM0_LOG_MAX_LOG_REC] = {};
+	struct m0_be_dtm0_log  *log                          = NULL;
+
+	rc = m0_dtm0_clk_src_init(&cs, M0_DTM0_CS_PHYS);
+	M0_UT_ASSERT(rc == 0);
+
+	rc = m0_be_dtm0_log_init(&log, &cs, true);
+	M0_UT_ASSERT(rc == 0);
+
+	for (i = 0; i < UT_DTM0_LOG_MAX_LOG_REC; ++i) {
+		rc = ut_dl_init(&txd[i], &buf[i], i);
+		M0_UT_ASSERT(rc == 0);
+	}
+
+	/* pa[0].st = EX pa[1].st = IP pa[2].st = IP, pyld = valid */
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[0], M0_DTPS_EXECUTED);
+	m0_mutex_lock(&log->dl_lock);
+	rc = m0_be_dtm0_log_update(log, NULL, &txd[0], &buf[0]);
+	M0_UT_ASSERT(rc == 0);
+	rec[0] = m0_be_dtm0_log_find(log, &txd[0].dtd_id);
+	M0_UT_ASSERT(rec[0]);
+	rc = ut_dl_verify_log_rec(rec[0], 0);
+	M0_UT_ASSERT(rc);
+
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[1], M0_DTPS_PERSISTENT);
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[2], M0_DTPS_PERSISTENT);
+	rc = m0_be_dtm0_log_update(log, NULL, &txd[0], &buf[0]);
+	M0_UT_ASSERT(rc == 0);
+	rec[0] = m0_be_dtm0_log_find(log, &txd[0].dtd_id);
+	M0_UT_ASSERT(rec[0]);
+	rc = ut_dl_verify_log_rec(rec[0], 0);
+	M0_UT_ASSERT(rc);
+
+	rc = m0_be_dtm0_log_prune(log, NULL, &txd[0].dtd_id);
+	M0_UT_ASSERT(rc == -EPROTO);
+
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[0], M0_DTPS_PERSISTENT);
+	rc = m0_be_dtm0_log_update(log, NULL, &txd[0], &buf[0]);
+	M0_UT_ASSERT(rc == 0);
+
+	rc = m0_be_dtm0_log_prune(log, NULL, &txd[0].dtd_id);
+	M0_UT_ASSERT(rc == 0);
+
+	rec[0] = m0_be_dtm0_log_find(log, &txd[0].dtd_id);
+	M0_UT_ASSERT(rec[0] == NULL);
+
+	rc = m0_be_dtm0_log_prune(log, NULL, &txd[0].dtd_id);
+	M0_UT_ASSERT(rc == -ENOENT);
+
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[0], M0_DTPS_INPROGRESS);
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[1], M0_DTPS_INPROGRESS);
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[2], M0_DTPS_PERSISTENT);
+	rc = m0_be_dtm0_log_update(log, NULL, &txd[0], &empty_buf);
+	M0_UT_ASSERT(rc == 0);
+	rec[0] = m0_be_dtm0_log_find(log, &txd[0].dtd_id);
+	M0_UT_ASSERT(rec[0]);
```

Comment:
&& rec[0].dtd_id == txd[0].dtd_id

Reply:
ut_dl_verify_log_rec(rec[0], 0) is verifying all the content inside record just after this line.